src/connector/src/sink/postgres.rs

```diff
@@ -0,0 +1,626 @@
+                }
+            }
+
+            // check that pk matches
+            {
+                let pg_pk_names = pg_table.pk_names();
+                let sink_pk_names = self
+                    .pk_indices
+                    .iter()
+                    .map(|i| &self.schema.fields()[*i].name)
+                    .collect::<HashSet<_>>();
+                if pg_pk_names.len() != sink_pk_names.len() {
+                    return Err(SinkError::Config(anyhow!(
+                    "Primary key mismatch: Postgres table has primary key on columns {:?}, but sink schema defines primary key on columns {:?}",
+                    pg_pk_names,
+                    sink_pk_names
+                )));
+                }
+                for name in pg_pk_names {
+                    if !sink_pk_names.contains(name) {
+                        return Err(SinkError::Config(anyhow!(
+                        "Primary key mismatch: Postgres table has primary key on column `{}`, but sink schema does not define it as a primary key",
+                        name
+                    )));
+                    }
+                }
+            }
+        }
+
+        Ok(())
+    }
+
+    async fn new_log_sinker(&self, writer_param: SinkWriterParam) -> Result<Self::LogSinker> {
+        Ok(PostgresSinkWriter::new(
+            self.config.clone(),
+            self.schema.clone(),
+            self.pk_indices.clone(),
+            self.is_append_only,
+        )
+        .await?
+        .into_log_sinker(SinkWriterMetrics::new(&writer_param)))
+    }
+}
+
+struct Buffer {
+    buffer: Vec<StreamChunk>,
+    size: usize,
+}
+
+impl Buffer {
+    fn new() -> Self {
+        Self {
+            buffer: Vec::new(),
+            size: 0,
+        }
+    }
+
+    fn push(&mut self, chunk: StreamChunk) -> usize {
+        self.size += chunk.cardinality();
+        self.buffer.push(chunk);
+        self.size
+    }
+
+    fn drain(&mut self) -> Vec<StreamChunk> {
+        self.size = 0;
+        std::mem::take(&mut self.buffer)
+    }
+}
+
+pub struct PostgresSinkWriter {
+    config: PostgresConfig,
+    pk_indices: Vec<usize>,
+    is_append_only: bool,
+    client: tokio_postgres::Client,
+    buffer: Buffer,
+    insert_statement: Statement,
+    delete_statement: Option<Statement>,
+    merge_statement: Option<Statement>,
+}
+
+impl PostgresSinkWriter {
+    async fn new(
+        config: PostgresConfig,
+        mut schema: Schema,
+        pk_indices: Vec<usize>,
+        is_append_only: bool,
+    ) -> Result<Self> {
+        let client = create_pg_client(
+            &config.user,
+            &config.password,
+            &config.host,
+            &config.port.to_string(),
+            &config.database,
+            &config.ssl_mode,
+            &config.ssl_root_cert,
+        )
+        .await?;
+
+        // Rewrite schema types for serialization
+        let schema_types = {
+            let pg_table = PostgresExternalTable::connect(
```

Comment:
We called this once inside `validate`, but `validate` just does a check, we can't store the `pg_table` obtained from there. We have to call it again, so we can get the `pg_table`, and from there get the pg_types. This type info will tell us how to handle our dynamic values. For instance, the same value can be serialized to varchar / uuid, depending on the type info.